sklearn/model_selection/_search.py

```diff
@@ -574,6 +599,39 @@ def inverse_transform(self, Xt):
         self._check_is_fitted("inverse_transform")
         return self.best_estimator_.inverse_transform(Xt)
 
+    def _generate_warm_start_groups(self, candidate_params):
```

Comment:
I feel like this logic is complicated enough to be it's own function + have some unit tests. This function would only need `candidate_params` and `use_warm_start` to be passed in.